CIPs/0010.md

```diff
@@ -0,0 +1,84 @@
+
+Celo stakeholders have enjoyed the ability to segment responsabilities into different keys as that allows for different custodying patterns. That being said, the current implementations comes with limitations around 1) what accounts and 2) which domains can be authorized. This CIP tries to generalize both.
+
+## Motivation
+
+The `Accounts.sol` smart contract houses a general account abstraction that currently services the following purposes:
+- Acts as the canonical account for the Proof-of-Stake system (aka LockedGold account)
+- Keeps track of authorized signer keys in the PoS system (validator, voter, and attestation signers)
+- Stores basic profile information such as a name and mapping to `dataEncryptionKey` and `walletAddress`
+- Entry into an account's off-chain metadata
+
+Over the past months, it has become noticeable that especially the authorization of signer keys by the "account key" is a very useful method to authenticate the "account owner" without having to unnecessarily bring the "account key" online which is the most privileged, and thus most important to secure. Example of such cases have been:
+
+- The signing of the off-chain data should ideally not happen with the account key. During the development of the `ReleaseGold` smart contract it has been found that smart contracts as accounts cannot sign metadata. As a workaround, verifying metadata has been adjusted to allow any authorized signer to sign the payload, which violates the intention behind having well scoped authorized signers in the first place.
+- As part of ODIS, accounts have to authenticate with the ODIS signers for rate-limiting purposes. To do that, the "account key" or the "authorized DEK" has to sign the request which also violates the intention of keeping keys that can move funds as cold as possible.
+- For other applications such as CIP8 or other down-the-road applications, specific keys are going to be useful. Additionally, the `dataEncryptionKey` seems out of place without useful segmentation ability, despite it being effectively also an authorized key (if not signer).
+
+Allowing the segmentation of domains for a Celo account (i.e. an account on `Accounts.sol`) gives Celo stakeholders more flexibility in their custody setup. It allows sensitive domains (like moving balances) to be only accessible by keys that are "cold", i.e. harder to compromise, while avoiding making less sensitive domains more user friendly. Expanding such segmentation aids the user experience, both of interacting with the Celo Core Contracts themselves (via authorized smart contracts), as well as dapps that currently would need to copy this authorization scheme to receive similar benefits for their users. While this CIP does not specify further applications, per-dapp-authorizations can become practical.
```

Comment:
A couple of things:
1. "while avoiding making less sensitive [...]" should be "while making less sensitive [...]"
2. Where it says "currently would need to copy this authorization scheme", it's not fully clear to me which scheme it's referring to.  Is it to the scheme already implemented through `Account`?  Perhaps a bit more background and explanation on this point (what they would have to do and why) would be helpful for people like me who don't know the ins and outs as well.  For example, I'm not clear on whether they would have to copy it to define new domains or also for making use of existing domains in their dapps.

Reply:
Thanks @oneeman for the read

For `currently would need to copy this authorization scheme`, I tried to say that without generalizing this scheme for other applications, dapps would have to copy `Accounts` to get the same benefit for their application.